### ISSUE TITLE
HasIndex definition leads to incorrect migrations

### ISSUE THREAD

**Issue by TobiasBreuer** [aspnet/EntityFramework6#586]
Definition of Index is getting messed up when defining a new composed index including a column that already had an index due to foreign key convention. But - The foreign key convention seems not to be the relevant part. An existing index seems to be enough to see the issue.

This issue is closely related to #575, #518 and #460 
Please also see this [stackoverflow question](https://stackoverflow.com/questions/26055140/ef-migrations-drops-index-when-adding-compsite-index) as an additional example for what is wrong.

### Steps to reproduce
I've created a small demonstrator to reproduce the issue - Please note that it does not make much sense semantically but instead is only used to show the point: 

[EFIndexIssueDemonstrator.zip](https://github.com/aspnet/EntityFramework6/files/2239856/EFIndexIssueDemonstrator.zip)

The solution contains two migrations, the initial one, and the faulty one after adding the index.

The relevant part of the initial migration is:
```c#
CreateTable(
    "dbo.Pages",
    c => new
        {
            Guid = c.Guid(nullable: false),
            Number = c.Int(nullable: false),
            FinishDate = c.DateTime(nullable: false),
            BookId = c.Guid(nullable: false),
        })
    .PrimaryKey(t => t.Guid)
    .ForeignKey("dbo.Books", t => t.BookId)
    .Index(t => t.BookId);
```
Note, that an index has been defined for 'BookId' due to the foreign key convention.

After that, a new index has been defined in the `PageConfiguration` class:

```c#
public class PageConfiguration : EntityTypeConfiguration<Page>
{
    public PageConfiguration()
    {
        HasKey(p => p.Guid);

        // Added Index after defining Initial Migration:
        // This combined index will lead to the index on BookId (created by convention due to foreign key relationship defined in BookConfiguration)
        // to be dropped.
        // Changing the order here has no effect on the result.
        HasIndex(p => new { p.FinishDate, p.BookId });
    }
}
```
When applying this configuration after the initial migration this leads to following 'Up' method of the next migration:
```c#
public override void Up()
{
    DropIndex("dbo.Pages", new[] { "BookId" });
    CreateIndex("dbo.Pages", new[] { "FinishDate", "BookId" });
}
```
No my first guess was, that defining an explicit index on a column with foreign key relationship might somehow disable the convention, so I thought that defining the original index again in the PageConfiguration would overcome this issue, but it really does not.

It depends on whether this explicit index creation (`HasIndex(p => p.BookId);`) is specifiied before, or after the combined one.
When specified before, the resulting migration is exactly the same as it would not have been added at all.
When specified after, the combined index definition the result will look like this:
```c#
public override void Up()
{
    DropIndex("dbo.Pages", new[] { "BookId" });
    CreateIndex("dbo.Pages", "FinishDate", name: "IX_FinishDate_BookId");
    CreateIndex("dbo.Pages", "BookId");
}
```
i.e. the combined index is not specified at all, and the index on the BookId will get dropped and re-created afterwards.

### Further technical details
EF version: 6.2
Database Provider: MS-Sql-Server 2014 / MySQL 5.7
Operating system: Window 7 Enterprise x64 
IDE: Visual Studio 2017 15.7.5

**Comment by chrisblock** [comment]
I haven't investigated much, but I think this is also be resolved by #534.

**Comment by ajcvickers** [comment]
Same root cause as #460, which is now fixed.
